### PR TITLE
Hello, Aya Native.

### PR DESCRIPTION
**Pull request by imkiva**
This PR integrates Aya with GraalVM's native image.

### CODE REVIEW

File: buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      .flatMap(line -> expand(line).stream())
+      .map(line -> generateConfig(line))
+      .collect(Collectors.joining(",\n", "[\n", "]\n"))
+    Files.writeString(outputDir.toPath().resolve("reflect-config.json"), lines)
+  }
+
+  def List<String> expand(String line) {
+    var matcher = pattern.matcher(line)
+    if (!matcher.find()) return List.of(line)
+    int start = java.lang.Integer.parseInt(matcher.group(1))
```
Comment (ice1000) [inline buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy:38]:
```suggestion
    int start = matcher.group(1) as int
```

File: buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    group = "build setup"
+  }
+
+  @OutputDirectory File outputDir
+  @InputFile File inputFile
+
+  private var pattern = Pattern.compile("\\{(\\d+)\\,\\s*(\\d+)\\}")
+
+  @TaskAction def run() {
+    var lines = Files.readAllLines(inputFile.toPath()).stream()
```
Comment (ice1000) [inline buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy:26]:
```suggestion
    var lines = Files.lines(inputFile.toPath())
```

File: cli/build.gradle.kts [hunk 3]
```diff
@@ -21,6 +21,10 @@ dependencies {
   // testImplementation("org.ice1000.jimgui", "fun", version = deps.getProperty("version.jimgui"))
 }
 
+plugins {
+  id("org.graalvm.buildtools.native") version "0.9.11"
```
Comment (ice1000) [inline cli/build.gradle.kts:25]: Does this make Aya depend on a graalvm JVM?

**Review by ice1000 — COMMENTED** [review]
🤔

File: cli/build.gradle.kts [hunk 4]
```diff
@@ -31,3 +35,38 @@ if (isMac) tasks.withType<JavaExec>().configureEach {
...
+      useFatJar.set(true)
+    }
+  }
+
+  binaries.configureEach {
+    javaLauncher.set(javaToolchains.launcherFor {
+      languageVersion.set(JavaLanguageVersion.of(17))
+      vendor.set(JvmVendorSpec.matching("GraalVM Community"))
+    })
+  }
```
Comment (imkiva) [inline cli/build.gradle.kts:66]: @ice1000

File: cli/build.gradle.kts (hunk 3 above)
Comment (imkiva) [reply cli/build.gradle.kts:25]: No. Only the command `gradle nativeCompile` requires a GraalVM installation, and the installation doesn't have to be the Gradle JVM. See new review comments.

File: buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy (hunk 1 above)
Comment (imkiva) [reply buildSrc/src/main/groovy/org/aya/gradle/GenerateReflectionConfigTask.groovy:38]: Ohhh, I was using `.toInt()` and got mysterious type error, thank you!

File: cli/build.gradle.kts (hunk 3 above)
Comment (imkiva) [reply cli/build.gradle.kts:25]: And our CI on this PR still passes, thus Aya is still buildable on a GraalVM-free machine.

File: cli/build.gradle.kts (hunk 4 above)
Comment (ice1000) [reply cli/build.gradle.kts:66]: If I don't need to run the ni task, do I still have to install graalvm?

File: cli/build.gradle.kts (hunk 3 above)
Comment (ice1000) [reply cli/build.gradle.kts:25]: Nice

**Review by ice1000 — APPROVED** [review]
bors d=imkiva

**Comment by imkiva** [comment]
https://www.graalvm.org/22.0/reference-manual/native-image/Agent/

**Comment by imkiva** [comment]
In case I forgot this url: https://repo.gradle.org/ui/native/jcenter/org/graalvm/buildtools/

**Comment by imkiva** [comment]
ONE LAST STEP: build native image in CI

**Comment by ice1000** [comment]
Is it possible to configure the path of graalvm by gradle.properties?

**Comment by imkiva** [comment]
Related: https://github.com/oracle/graal/issues/4502

**Comment by imkiva** [comment]
> Is it possible to configure the path of graalvm by gradle.properties?

Did you mean the dependency in `buildSrc`?

**Comment by imkiva** [comment]
> Is it possible to configure the path of graalvm by gradle.properties?

Oh sorry, I got it. Currently no

**Review by ice1000 — APPROVED** [review]
I trust your code.

**Comment by imkiva** [comment]
> I trust your code.




Thank you~ then I will
bors merge

**Pull request merged**